### PR TITLE
[json] correctly store/restore `TList`

### PR DESCRIPTION
**Pull request by linev**
There is principal difference between `lst->Add(obj)` and `lst->Add(obj, "")`
In first case `TObjLink` is created, in second `TObjOptLink`.
When such `TList` was stored in JSON and read back, all entries in the `TList` were `TObjOptLink`
Now such situation marked specially and reconstructed `TList` object is correct.

Modify storage and reading part.
Add correspondent test.

IMPORTANT!!!

Normal ROOT I/O also has problem with such `TList`. When read from the buffer, 
all entries with **empty** options converted into `TObjLink`. But `TObjLink::GetOption()` returns not empty string 
but `obj->GetOption()`. 

This can change drawing of TCanvas, when such canvas read from the file. 
There are many classes like `TH1` or `TPave` implementing `GetOption()` method and 
potentially returning non-empty string.

### CODE REVIEW

**Comment by linev** [comment]
Here modified test which fails with normal I/O:
```
TEST(TMemFile, TList_Add_Without_Option)
{
   TList lst0;
   lst0.Add(new TNamed("name0", "title0"));  // without option
   lst0.Add(new TNamed("name1", "title1"), ""); // with empty option
   lst0.Add(new TNamed("name2", "title2"), "option2"); // with non-empty option

   auto f = new TMemFile("test.root", "RECREATE");
   f->WriteTObject(&lst0, "list");

   lst0.Delete();

   auto lst1 = f->Get<TList>("list");
   delete f;

   EXPECT_NE(lst1, nullptr);

   auto link = lst1->FirstLink();

   EXPECT_STREQ("name0", link->GetObject()->GetName());
   EXPECT_STREQ("title0", link->GetObject()->GetTitle());
   EXPECT_STREQ("", link->GetAddOption()); // by default empty string returned
   EXPECT_EQ(dynamic_cast<TObjOptLink *>(link), nullptr);

   link = link->Next();

   EXPECT_STREQ("name1", link->GetObject()->GetName());
   EXPECT_STREQ("title1", link->GetObject()->GetTitle());
   EXPECT_STREQ("", link->GetAddOption());
   EXPECT_NE(dynamic_cast<TObjOptLink *>(link), nullptr); // this will fail in normal ROOT I/O

   link = link->Next();

   EXPECT_STREQ("name2", link->GetObject()->GetName());
   EXPECT_STREQ("title2", link->GetObject()->GetTitle());
   EXPECT_STREQ("option2", link->GetAddOption());
   EXPECT_NE(dynamic_cast<TObjOptLink *>(link), nullptr);

   link = link->Next();

   EXPECT_EQ(link, nullptr);
}
```

File: io/io/src/TBufferJSON.cxx [hunk 1]
```diff
@@ -1610,61 +1610,82 @@ void TBufferJSON::JsonWriteCollection(TCollection *col, const TClass *)
    // collection treated as JS Array
    AppendOutput("[");
 
-   bool islist = col->InheritsFrom(TList::Class());
    TMap *map = nullptr;
-   if (col->InheritsFrom(TMap::Class()))
+   TList *lst = nullptr;
+   if (col->InheritsFrom(TList::Class()))
+      lst = dynamic_cast<TList *>(col);
+   else if (col->InheritsFrom(TMap::Class()))
```
Comment (pcanal) [inline io/io/src/TBufferJSON.cxx:1617]: Apriori we can simply the code with:
```suggestion
   TList *lst = dynamic_cast<TList *>(col);
   TMap *map = dynamic_cast<TMap*>(col);
```
i.e. I don't think that `InheritsFrom` is much faster than the `dynamic_cast`.

Comment (linev) [reply io/io/src/TBufferJSON.cxx]: I add commit with such changes

**Review by pcanal — APPROVED** [review]
LGTM.  Please open a new issue for the related binary I/O issue.

**Pull request merged**